Add texlive

ocefpaf opened the pull request:
@janschulz and @jakirkham do you want to be added here as maintainers?

Closes #359

conda-forge-linter commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/texlive-core```) and found it was in an excellent condition.

janschulz commented:
@ocefpaf No, but I will subscribe to the repo :-)

jakirkham commented on recipes/texlive-core/build.sh:
```diff
@@ -0,0 +1,52 @@
+               --with-libpng=$PREFIX \
+               --with-zlib=$PREFIX \
+               --x-libraries=$PREFIX/lib \
+               --x-includes=$PREFIX/include \
```
Can't assume X11 on Mac as it is no longer shipped by default. Also, I don't think X11 will live there (unless we build it 😄). Maybe we should just drop these.

jakirkham replied:
While that sounds crazy it isn't entirely, see this [recipe]( https://github.com/ostrokach/conda-recipes-extra/tree/8ea3eab3de34f1ee043a7df3c4a24e44b9e48e18/xorg ).

ocefpaf replied:
Or https://github.com/pkgw/conda-recipes/tree/master/recipes/modern-xorg-stack

jakirkham replied:
Yep, would be great if we could @pkgw to join the effort somehow. 😉

jakirkham commented:
This test `pdftexdir/ttf2afm.test` is failing on Travis CI. The rest are fine.

jakirkham commented on recipes/texlive-core/build.sh:
```diff
@@ -0,0 +1,52 @@
+               --without-system-poppler
+
+  make
+  make check
```
Could we please change this to `make check || cat test-suite.log`? That should at least tell us about the test failure on Travis CI.

jakirkham replied:
Ignore that the logs are in each library directory. The issue is during "Web2C 2015" test suite. Here is what I found in `$SRC_DIR/tmp_build/texk/web2c/test-suite.log`.

```
==================================
   Web2C 2015: ./test-suite.log
==================================

# TOTAL: 69
# PASS:  68
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: pdftexdir/ttf2afm
=======================

*** ttf2afm postV3.ttf
sed: RE error: illegal byte sequence
*** ttf2afm postV7.ttf

Warning: ttf2afm (file postV7.ttf): unsupported format (00070000) of `post' table, assuming 3.0

Warning: ttf2afm (file postV7.ttf): no names available in `post' table, print glyph names as indices
sed: RE error: illegal byte sequence

failed tests: postV3 postV7

```

jakirkham replied:
Based on this error message my recommendation is to change `make check` -> `LC_ALL=C make check`.

jakirkham replied:
Yes, I checked locally. If you add `LC_ALL=C ` before `make check`, the tests pass on Mac. The problem is best explained by this [stackoverflow answer]( http://stackoverflow.com/a/23584470 ).

jakirkham commented on recipes/texlive-core/build.sh:
```diff
@@ -0,0 +1,52 @@
+               --with-pixman=$PREFIX \
+               --with-freetype2=$PREFIX \
+               --with-libpng=$PREFIX \
+               --with-zlib=$PREFIX \
```
I'm a little worried that these `--with-*` lines are not doing what we think. It looks like it builds its own copies of these. We need to change this. Try downloading the source and look in the `libs` directory, I think you will see what I mean.

jakirkham replied:
That or these lines are being ignored.

jakirkham commented on recipes/texlive-core/meta.yaml:
```diff
@@ -0,0 +1,48 @@
+        - icu
+        - mpfr
+        - pixman
+        - freetype 2.5.*
```
Should add `harfbuzz`.

ocefpaf replied:
Actually that won't work. We need to use the bundled version of the lib.

jakirkham commented on recipes/texlive-core/build.sh:
```diff
@@ -0,0 +1,52 @@
+               --with-zlib=$PREFIX \
+               --x-libraries=$PREFIX/lib \
+               --x-includes=$PREFIX/include \
+               --without-system-harfbuzz \
```
Would be nice if we could get this. There is a recipe in conda-recipes.

ocefpaf replied:
https://github.com/pkgw/conda-recipes/tree/a03809015b67b174dc978eb4d0e978a2c0e3b8da/recipes/harfbuzz

jakirkham replied:
I believe it is also in conda-recipes.

ocefpaf commented:
Thanks for the review @jakirkham! I guess that the remaining commits I had here address most of your questions.

These are the bundled libraries we are using right now.

```
> cat log.txt | grep "library from TL"
configure: Using `harfbuzz' headers and library from TL tree
configure: Assuming `teckit' headers and library from TL tree
configure: Using `graphite2' headers and library from TL tree
configure: Assuming `zziplib' headers and library from TL tree
configure: Assuming `xpdf' headers and library from TL tree
configure: Using `poppler' headers and library from TL tree
configure: Assuming `mpfr' headers and library from TL tree
configure: Using `cairo' headers and library from TL tree
configure: Assuming `potrace' headers and library from TL tree
configure: Assuming `gd' headers and library from TL tree
configure: Assuming `libpaper' headers and library from TL tree
configure: Assuming `ptexenc' headers and library from TL tree
configure: Assuming `kpathsea' headers and library from TL tree
```

We don't have `graphite2` and, even though the default channel has `harfbuzz`, `harfbuzz` won't work without `graphite2`.

We do have `cairo`, but the recipe breaks and TL cannot find it with I try to specify `cairo` with:

```
--with-system-cairo

# and/or

--with-cairo-includes=$PREFIX/include --with-cairo-libdir=$PREFIX/lib
```


The same happens for `mpfr`, but at least the recipe does not break and the bundle lib kicks in.

ccordoba12 commented:
Wow, this is huge guys!! Thanks for working on it @ocefpaf :-)

Question: How is the installation of new Tex packages going to be handled? Is Texlive package manager going to be part of the installation? (I think Texlive has one, as Miktex, right?)

jakirkham commented on recipes/texlive-core/meta.yaml:
```diff
@@ -0,0 +1,51 @@
+        - icu
+        - mpfr
+        - pixman
+        - harfbuzz
```
Apparently this is Linux only. Let's add a selector so we can deal with that until this changes.

ocefpaf replied:
> Wow, this is huge guys!! Thanks for working on it @ocefpaf :-)

Want to be a maintainer :wink: 

> Question: How is the installation of new Tex packages going to be handled? Is Texlive package manager going to be part of the installation? (I think Texlive has one, as Miktex, right?)

That is a good question! But I don't have the answer. I am interested into having some basic Tex capabilities in a conda env. It would be nice to find more maintainers for this recipe.

jakirkham commented:
Linux can't find `zlib`. Not sure why. It seems to find most of the other dependencies.

pkgw commented:
Hmmm, I can't find the comment where I got CC'ed on this, but this PR certainly of interest to me. Did anyone have any particular questions they wanted to ask about [the recipe I put together](https://github.com/pkgw/conda-recipes/tree/master/recipes/texlive-core)?

I am also curious about the handling of individual Texlive packages. Packaging them all separately seems extremely painful but I'll fully admit that [my approach](https://github.com/pkgw/conda-recipes/blob/master/recipes/texlive-selected/make-tarball.sh) is far from elegant.

ocefpaf commented:
> Wow, this is huge guys!! Thanks for working on it @ocefpaf :-)

Want to be a maintainer :wink: 

> Question: How is the installation of new Tex packages going to be handled? Is Texlive package manager going to be part of the installation? (I think Texlive has one, as Miktex, right?)

That is a good question! But I don't have the answer. I am interested into having some basic Tex capabilities in a conda env. It would be nice to find more maintainers for this recipe.

jakirkham commented:
> Hmmm, I can't find the comment where I got CC'ed on this, but this PR certainly of interest to me.

Ah, good, I hoped we would get your attention. 😄 

> Did anyone have any particular questions they wanted to ask about the recipe I put together?

Yeah, we were actually looking at your xorg recipe. Would be pretty neat to get that rolling here in a new PR if you would be interested.

> I am also curious about the handling of individual Texlive packages. Packaging them all separately seems extremely painful but I'll fully admit that my approach is far from elegant.

Yeah, this is a good question. Imagine at some point automation is going to part of the solution here for the near term. Personally, am thinking we get this out there and see how people use it a bit. Though we might want to play with some sort of `conda skeleton` solution for quickly generating recipes for this for this sort of stuff.

ocefpaf commented:
> Hmmm, I can't find the comment where I got CC'ed on this, but this PR certainly of interest to me. Did anyone have any particular questions they wanted to ask about [the recipe I put together](https://github.com/pkgw/conda-recipes/tree/master/recipes/texlive-core)?

Awesome that you are interested @pkgw! @jakirkham pointed me to your recipe, but I have no idea what I am doing here :smile:  Any input is welcomed!

BTW would you be interested in being a maintainer of this recipe here at conda-forge?  

> I am also curious about the handling of individual Texlive packages. Packaging them all separately seems extremely painful but I'll fully admit that [my approach](https://github.com/pkgw/conda-recipes/blob/master/recipes/texlive-selected/make-tarball.sh) is far from elegant.

I am not working on that right now, but we should think more about it. I agree that it is not elegant, but it works.

janschulz commented:
> Question: How is the installation of new Tex packages going to be handled? Is Texlive package manager going to be part of the installation? (I think Texlive has one, as Miktex, right?)

package manager: https://www.tug.org/texlive/doc/tlmgr.html

IMO: install the most important packages into this package (e.g. what nbconvert uses and some more which everybody seems to use - no idea where to get such a list? Debian -> https://packages.debian.org/sid/texlive-latex-recommended) and then setup the texlive install so that it uses the packages in the conda env as base and installs new one in a user location (e.g. this: https://www.tug.org/texlive/doc/tlmgr.html#USER-MODE?).

jakirkham commented:
Restarted CI as it looked like some Cython stuff got stuck in here. Want to get a clean build log so we can see what is going on.

ivoflipse commented:
We're also interested in a Tex package for Linux so you can add me, @Korijn and @maxyme as maintainers. 

Is there anything we can do to help this recipe move forward?

ocefpaf commented:
@ivoflipse I can either remove the OS X now and we open an issue to address it in the feedstock or you can take over and try to fix OS X. (I don't own a Mac to debug the problem there.)

ivoflipse commented:
@ocefpaf None of us work on OS X regularly either, so I'm fine with fixing that later on. I'm mostly interested in a Linux build

Korijn commented on recipes/texlive-core/meta.yaml:
```diff
@@ -0,0 +1,49 @@
+
+test:
+    commands:
+        - pdftex --version
```
Maybe add a call to bibtex, and perhaps there is a package manager that we can test as well?

ocefpaf commented:
I removed `OS X` for now and added a simple test for `bibtex`. Perl is needed for testing, but I am unsure if it is a run time dependency too.

I am merging this as is and addressing the renaming issue in the feedstock.
